### ISSUE TITLE
Real time combat

### ISSUE THREAD

**Issue by blueFireGames** [rpgboss/rpgboss#132]
It would be nice if there was an option to use real time combat (similar to Zelda) in addition to the current battle system.  This would allow the creator to create some unique encounters (such as two monsters coming from one direction while a third simultaneously attacks from another).

Jason

**Comment by devnewton** [comment]
Hello,

I am the author of nanimstudio, a sprite animation software in Java. It has a json import/export that looks a little like the one used by rpgboss.

http://devnewton.bci.im/softwares/nanim

If you are interested, I can try to build a swing component based on nanimstudio so you will be able to integrate sprite animation editing in rpgboss.

**Comment by blueFireGames** [comment]
I think this would be good.

Jason

**Comment by tommycli** [comment]
devnewton: Yes, please do.

Try to figure out what would be ideally needed for our project.

Currently all that exists is the SpriteSet resource file which follows this fixed format: http://s444.photobucket.com/user/grandmadeb_rmvx/media/xmq0_chara1.png.html

As you can see it only supports 'walking'.

Could you conceptualize an animation format that's both easy to create / edit and supports walking, running, attacking, being hit, etc.?

I am open to adopting the JSON format used by your software if you are willing to provide long term support.

**Comment by devnewton** [comment]
Hello,

We can say that I provide "long term support" because my format is used in my games :-)

- [Newton Adventure](http://newtonadventure.bci.im).
- [Ned et les maki](http://en.geekygoblin.org/ned-et-les-maki/).

Here some sample:

- A platformer character: [json](https://github.com/devnewton/newton_adventure/blob/master/game/data/default_level_data/hero.json) with one [image](https://github.com/devnewton/newton_adventure/blob/master/game/data/default_level_data/hero.png) including all frames created using texture packing optimization built in nanimstudio software.
- A shining key with separated images: [json](https://github.com/devnewton/newton_adventure/blob/master/game/data/default_level_data/key.json). Images: [0](https://github.com/devnewton/newton_adventure/blob/master/game/data/default_level_data/key-0.png), [1](https://github.com/devnewton/newton_adventure/blob/master/game/data/default_level_data/key-1.png), [2](https://github.com/devnewton/newton_adventure/blob/master/game/data/default_level_data/key-2.png), [3](https://github.com/devnewton/newton_adventure/blob/master/game/data/default_level_data/key-3.png), [4](https://github.com/devnewton/newton_adventure/blob/master/game/data/default_level_data/key-4.png), [5](https://github.com/devnewton/newton_adventure/blob/master/game/data/default_level_data/key-5.png), [6](https://github.com/devnewton/newton_adventure/blob/master/game/data/default_level_data/key-6.png), [7](https://github.com/devnewton/newton_adventure/blob/master/game/data/default_level_data/key-8.png), [0](https://github.com/devnewton/newton_adventure/blob/master/game/data/default_level_data/key-8.png).

Do you think it will suit your needs?

**Comment by tommycli** [comment]
Yeah, I can get behind that animation format. It seems sensible and general purpose.

I have a few question

- Is the duration in milliseconds? We've opted to use all seconds so far.

- I assume the u v coords are specified in terms of percentage of image? (1.0, 1.0) is bottom right?

- We already have an animation format I hacked together for use with attacks in battles. https://github.com/rpgboss/rpgboss/blob/master/common/src/main/resources/defaultrc/animations.json  How would one integrate specifying sound effects that happen at certain time points in your animation?

Thanks!

**Comment by tommycli** [comment]
Okay, getting down to more brass tacks:

Our current animations are comprised of two kinds.
- Basic walking. The animations are of a fixed set of frames and u v coords. See here as an example: https://github.com/rpgboss/rpgboss/blob/master/common/src/main/resources/defaultrc/spriteset/sys/vx_chara01_b.png 
  - Would your format be able to support this? Perhaps an import feature that automatically takes in the image, and generates the appropriate walk animations?
 - Battle animations, as I mentioned in the above comment. My main concern here is being able to integrate sounds at specific times.

It would be awesome if we could somehow have a unified and general-purpose 2d animation specification format that could meet our needs.

**Comment by devnewton** [comment]
> Is the duration in milliseconds?

Yes.

> I assume the u v coords are specified in terms of percentage of image?

It is the opengl texture coordinates system.

> I assume the u v coords are specified in terms of percentage of image? (1.0, 1.0) is bottom right?

Yes. I use 0-1.0 range instead to be able to resize an image in an external drawing program without having to change the json animation description.

> How would one integrate specifying sound effects that happen at certain time points in your animation?

I would like to integrate sound effects, hit boxes or transforms (translate, rotate, scale), but this needs some thoughts, because maybe it is beyond the scope of sprite animation. There are some existing tools available for this (Pencil2D, Synfig studio), but I've always found them overkill for oldschool 2d games.

Maybe we can set up a meeting to discuss what we want?

**Comment by tommycli** [comment]
Sounds good.

Send me an email at rpgboss@keepword.com.

**Comment by HeadClot** [comment]
Hey has there been any movement on this feature request?

Just curious :)
